### DISCUSSION TITLE
Device test runners for mobile devices (and .NET MAUI)

### DISCUSSION THREAD

**Discussion by mattleibow** [xunit/xunit#2739] [Ideas]
Hi folks, I am not sure what the plans are for the 3.x for mobile devices. I know for 2.x we had some repos:

 - https://github.com/xunit/devices.xunit
 - https://github.com/xunit/uitest.xunit

This was useful when running tests for mobile and I used it a fair bit.

While developing for .NET MAUI I took all that good code and made it work as well as did some new things. I have collected it all and put it as a standalone thing in my personal repo: 

https://github.com/mattleibow/Xunit.Runner.Devices

I am still using the 2.x APIs and I am using .NET 7, but so far it all works. I have also added support for XHarness:

https://github.com/dotnet/xharness

All the projects are split so you technically don't _have_ to use .NET MAUI. But for now, I have just made .NET MAUI test runners work (you don't need to have a MAUI app to use the runner, I just wrote the UI in MAUI). I have not yet finished the framework-free/native-only versions of the runners, but since you can test non-MAUI code it is a bit lower on the priority.

I have no real readme yet and all my commands are in my terminal history. 😄 

I have tested Android, iOS and Mac Catalyst apps both using a visual runner with a debugger as well as using XHarness.

**Comment by bradwilson** [comment]
There are no current plans for first party support for MAUI. The owner of those repos has decided to step away from xUnit.net. I have none of (a) experience, (b) time, nor (c) desire to own device support.

It's good to see someone from the community step up to provide these! Thanks a bunch. 😄 🎉 

(Out of curiosity: Have you tried your projects with 2.5 yet? The current prerelease build on NuGet is very likely to become the 2.5 final version, so it would be great if you could verify that your stuff works with 2.5 and/or be able to file any issues related to it.)

**Reply by mattleibow** [reply]
So far the 2.5 builds seem OK, I will continue to test them and see what happens!

**Reply by mattleibow** [reply]
> It's good to see someone from the community step up to provide these! Thanks a bunch. 😄 🎉 


I am trying to move my repo into the CommunityToolkit org so it can be community managed and maintained. They have a preferred license of MIT although they might also be OK with Apache 2.0... Just wondering if you would be open to adding the MIT license as well so it is all a single license? No _need_ but it will be very nice to all be MIT.

**Reply by bradwilson** [reply]
I don't know what's involved in a license change on my side, since that would have to be approved by .NET Foundation. MIT and Apache are compatible licenses, and either way you'd have to include an additional copyright/license statement since the code owners would be different.
